### PR TITLE
Wonderful Thor Hack

### PR DESCRIPTION
**Pull request by dwbuiten**
See comments for details on how the constants were derived, and the algorithm in general.

The intent is that this is used as the fast path once we add trellis quantization.

This particulrily helps high motion clips like Red Kayak and Ducks Take Off:

|                          Video |    PSNR | PSNR HVS |    SSIM | CIEDE 2000 | MS SSIM |    VMAF |
|                           ---: |    ---: |     ---: |    ---: |       ---: |    ---: |    ---: |
| ducks_take_off_1080p50_60f.y4m | -3.5880 |  -4.6035 | -3.3884 |    -1.3255 | -4.1065 | -9.5272 |
|         red_kayak_360p_60f.y4m |  0.1735 |  -2.2590 | -0.4001 |    -1.7106 | -1.9463 | -9.1934 |

AWCY Averages (varies a lot more by clip):

|    PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000 |   VMAF |
|    ---: |    ---: |    ---: |     ---: |    ---: |    ---: |       ---: |    --: |
| -0.2979 | -3.7618 | -2.8471 |  -0.5794 | -0.2043 | -0.3993 |    -1.5294 | -3.300 |

AWCY Link: https://beta.arewecompressedyet.com/?job=master-162ee92ca6a819c19c6053841a4ab754cb735b16&job=deadzoning1%402019-09-02T14%3A31%3A02.785Z

Implements #1338 and #1340.

### CODE REVIEW

File: src/quantize.rs [hunk 1]
```diff
@@ -212,10 +214,40 @@ impl QuantizationContext {
     self.ac_quant = ac_q(qindex, ac_delta_q, bit_depth) as u32;
     self.ac_mul_add = divu_gen(self.ac_quant);
 
+    // All of these biases were derived by measuring the cost of coding
+    // a zero vs coding a one any any given coefficient position, or, in
+    // the case of end-of-block bias, the cost of coding the block with
+    // the eob was chosen (rounding to one) vs rounding to zero and continuing
+    // to choose a new eob. This was done over several clips, with the
+    // average of the bit costs taken over all blocks in the set, and a new
+    // biased derived via the method outlined in Jean-Marc Valin's
```
Comment (tmatth) [inline src/quantize.rs]: nit: should be `bias derived`

File: src/quantize.rs [hunk 2]
```diff
@@ -212,10 +214,40 @@ impl QuantizationContext {
...
+    // threshold = 0.5 + (lambda * avg_rate_diff) / 2.0
+    // bias = 1 - threshold
+    //
+    // lambda is a constant since our offsets are already adjusted for the
+    // quantizer.
+    //
+    // Biases were then updated, and cost collection was re-run, until
+    // the calculated biases started to converge after 2-4 iterations.
+    //
+    // In theory, the rounding biases for inter should be somwhat smaller
```
Comment (tmatth) [inline src/quantize.rs]: nit: somewhat

File: src/quantize.rs [hunk 3]
```diff
@@ -212,10 +214,40 @@ impl QuantizationContext {
...
+    // than the biases for intra, but this turns out to only be the case
+    // for eob optimization, or at least, is covered by eob optimization.
+    // The RD-optimal rounding biases for the actual coefficients seems
```
Comment (tmatth) [inline src/quantize.rs]: nit: should be "seem to be quite clos(ed?)"

File: src/quantize.rs [hunk 4]
```diff
@@ -224,20 +256,60 @@ impl QuantizationContext {
...
+    }
+
+    // We skip the DC coefficient since it has its own quantizer index.
+    let last_pos = if is_zero { pos } else { pos + 1 };
+
     qcoeffs[0] = coeffs[0] << (self.log_tx_scale as usize);
     qcoeffs[0] += qcoeffs[0].signum() * T::cast_from(self.dc_offset);
     qcoeffs[0] = T::cast_from(divu_pair(qcoeffs[0].as_(), self.dc_mul_add));
 
+    // Here we uses different rounding biases depending on whether we've
```
Comment (tmatth) [inline src/quantize.rs]: nit: `we use`

File: src/quantize.rs [hunk 5]
```diff
@@ -224,20 +256,60 @@ impl QuantizationContext {
...
+    // had recent coefficients that are larger than one, or less than
+    // one. The reason for this is that a block usually has a chunk of
+    // large coefficients and a tail of zeroes and ones, and the tradeoffs
+    // for coding these two are different. In the tail of zeroes and ones,
+    // you'll likely end up spending most bits just saying where that
+    // coefficient is in the block, whereas the in the chunk of larger
```
Comment (tmatth) [inline src/quantize.rs]: nit: `whereas in the`

File: src/quantize.rs [hunk 6]
```diff
@@ -212,10 +214,40 @@ impl QuantizationContext {
     self.ac_quant = ac_q(qindex, ac_delta_q, bit_depth) as u32;
     self.ac_mul_add = divu_gen(self.ac_quant);
 
+    // All of these biases were derived by measuring the cost of coding
+    // a zero vs coding a one any any given coefficient position, or, in
+    // the case of end-of-block bias, the cost of coding the block with
+    // the eob was chosen (rounding to one) vs rounding to zero and continuing
```
[the remark this answers is not in the document]
Comment (barrbrain) [reply src/quantize.rs]: alternate:  `the chosen eob`; my reading of the original was `the eob that was chosen`.

**Comment by dwbuiten** [comment]
@rzumer done

**Comment by dwbuiten** [comment]
Also approved by Tim on IRC.

**Pull request merged**